[#2697] improvement(core): Support different entity with the same name under a namespace

yuqi1129 opened the pull request:
### What changes were proposed in this pull request?

Support store entities with the same name under a namespace. 

### Why are the changes needed?

It's a necessary feature for Gravitino. 

Fix: #2697 

### Does this PR introduce _any_ user-facing change?

N/A.

### How was this patch tested?

UTs

jerryshao commented on core/src/main/java/com/datastrato/gravitino/Entity.java:
```diff
@@ -41,6 +43,35 @@ public static EntityType fromShortName(String shortName) {
+        case COLUMN:
+          return ImmutableList.of(METALAKE, CATALOG, SCHEMA, TABLE);
+        default:
+          throw new IllegalArgumentException("Unknown entity type: " + entityType);
```
Can you please rebase your PR based on @qqqttt123 newly added USER and GROUP support.

jerryshao commented:
I think the key thing here is to make this change backward compatible, otherwise there's no way for users to upgrade.

yuqi1129 commented:
> I think the key thing here is to make this change backward compatible, otherwise there's no way for users to upgrade.

Yeah, that is why I'm hesitant about it.

jerryshao commented on core/src/main/java/com/datastrato/gravitino/storage/kv/KvEntityStore.java:
```diff
@@ -199,72 +189,15 @@ public <E extends Entity & HasIdentifier> E update(
           nameMappingService.updateName(
-              generateKeyForMapping(ident), generateKeyForMapping(updatedE.nameIdentifier()));
+              generateKeyForMapping(ident, entityType, nameMappingService),
+              generateKeyForMapping(updatedE.nameIdentifier(), entityType, nameMappingService));
```
Only changing here is enough, do you need to change put/get also?

yuqi1129 replied:
all have been changed through method `concatIdAndName` and `put/get/update` will use this method to generate key format.

yuqi1129 commented on core/src/main/java/com/datastrato/gravitino/Entity.java:
```diff
@@ -43,6 +45,37 @@ public static EntityType fromShortName(String shortName) {
+          return ImmutableList.of();
+        case CATALOG:
+        case USER:
+        case GROUP:
```
@qqqttt123 , Can you confirm if these changes are acceptable?

qqqttt123 replied:
Now, My nameIdentifier is `metalake."system"."user".user1` and `metalake."system"."group".group1`.

yuqi1129 commented:
> > I think the key thing here is to make this change backward compatible, otherwise there's no way for users to upgrade.
> 
> Yeah, that is why I'm hesitant about it.

I have updated the code and added support for backward compatibility. Please take a look if you have time. Thanks.

jerryshao commented on core/src/main/java/com/datastrato/gravitino/storage/kv/BinaryEntityEncoderUtil.java:
```diff
@@ -0,0 +1,185 @@
+        prefixes.add(replacePrefixTypeInfo(encode, FILESET.getShortName()));
+        break;
+      case TABLE:
+      case FILESET:
```
Should you also add USER and GROUP here?

qqqttt123 replied:
If we add USER and GROUP here, USER and GROUP will block the non-cascade deletion of metalake.

jerryshao commented on core/src/main/java/com/datastrato/gravitino/storage/kv/BinaryEntityEncoderUtil.java:
```diff
@@ -0,0 +1,185 @@
+                Arrays.stream(namespaceIds).mapToObj(String::valueOf).collect(Collectors.toList()));
+    // We need to make it backward compatible, so we need to check if the name is already prefixed.
+    String mappingName =
+        VERSION_0_4_0_COMPATIBLE_ENTITY_TYPES.contains(type)
```
Better to change to `VERSION_0_4_COMPATIBLE_ENTITY_TYPES` to ignore the minor version.

jerryshao replied:
@yuqi1129 please fix this issue, thanks!

jerryshao reviewed (approved):
LGTM. Thanks for your work @yuqi1129
